### DISCUSSION TITLE
Should I 'grow' types mapped to entities with custom fields or move this logic to the controller as a separate query?

### DISCUSSION THREAD

**Discussion by StefanGramadnikov** [thecodingmachine/graphqlite#633] [Q&A]
I'm building a new structure and I'm not sure what is the correct way of handling things. 

I have three tables: 

```
challenge: 
    id,
    name,
    start_date,
    end_date
```

```
challenge_participant
    challenge_id,
    user_id,
    activities_type_1_count,
    activities_type_2_count,
```

```
challenge_prize
    name,
    challenge_id,
    activities_attended_required,
    prize
```

Based on the three tables my FE needs to pull the following data:

1. active challenges
2. challenge participant current progress (activities_type_1_count & activities_type_2_count)
3. leader board -> top 5 performing participants
4. prizes won for given participant

What I want to build is three types classes which will be exactly mapped to my DB structure, so I will have:

ChallengeType with fields `name`, `start_date`, `end_date` & `prizes ([PrizeType])` (prizes will come trough the relationship)
ChallengeParticipant with fields  `activities_type_1_count`, `activities_type_2_count` & `UserType` (again coming from relationship)
ChallengePrizeType with fields `activities_attended_required` & `prize`

Then on the controller level I would have three queries 
1. query `activeChallenges` -> returns `[ChallengeType]`
2. query `challengeParticipant` -> returns `ChallengeParticipantType`
3. query `challengeLeaderboard` -> aggregates the data and returns array of `[ChallengeParticipantType]`

The benefits I see in this from back-end prespective are:
1. I will have slim `Type classes`, which will be exactly mapped to the `Symfony entities`, without any additional custom fields on them
2. I will use services for calculating the leaderboard which will be injected in the controller, this way the controller will handle the request, the service will handle the business logic and return correct array of types. 
3. FE can still ask for the data from the three queries at once if they want. 

What my FE team suggests would be better from their prespective:

Create a `ChallengeType` with the following fields 
```
name: string! # unique string
startDate: DATE_TIME!
endDate: DATE_TIME!
prizes: [ChallengePrizeType!]!
# Custom fields (outside of the entity scope):
leaderboard: [ChallengeParticipantType!]!
currentStudentProgress: ChallengeParticipantType | null
```

What I don't like in this approach is that I would have one bigger type and I will have to put logic for calculating the leader board and fetching the `currentStudentProgress` inside the `ChallengeType`. 

Additionally if we need more custom type fields, it will grow even more. 

Essentially the question here is: `what is the better solution for putting custom logic` 

A. Should I keep my models slim and related to my Symfony entities and move the custom logic as additional GQL queries (controller level) or if the custom fields get too much, create a new totally detached custom type (for example `ChallengeReportType`)
B. Should I freely add `custom fields` into the GQL type is most suitable for me
C. Is there some other/better options I'm missing ?

I really don't like the second approach, because in the long term, stuffing logic into those types always seems like a bad idea. I really like to keep things as simple as possible and divide the responsibility as much as possible. 
For example: 
Controller should handle the request (validation, security, routing) and return the correctly mapped response, a service shold handle the business logic, and the Type classes should be ONLY for mapping data, nothing more. 

I tried to find an answer to my question by digging trough different kind of GQL docs and articles, but nothing really gave my any helpful info. 
I would really appreciate some input on this! Thanks in advance!

**Comment by oojacoboo** [comment]
Hey, so there is the `ExtendType` attribute that is excellent for extending types to keep logic out of your main domain models, also frequently your `Type` annotated classes.  This is great for adding specific fields that need access to services, custom queries or calculations, etc.  Another option, is to use the `Autowire` attribute that you can use to autowrire a service into a model method, as an argument.  However, I prefer the `ExtendType`, myself, as I find that to be a cleaner separation of concerns.

Let me know if that helps.
